### PR TITLE
feature: add write back for memory cache

### PR DESCRIPTION
**Pull request by hookokoko**
feature: add write back for memory cache

### CODE REVIEW

**Comment by hookokoko** [comment]
> If we want to create a general solution, we need to think about define the OnEvict for all implementations. It's a little bit difficult

（英语渣担心描述不清楚，还是中文了）
核心问题是对于接入的cache，如果底层不支持过期回调的API，在不修改源码的情况下没办法做到。
除非，我们在外围也就是writeback本身添加过期删除策略。因为无法在key过期的时候主动通知我们，所以只能是我们主动询问。

参考redis的过期删除策略：
1. 惰性删除：这种情况差不多就变成了read though场景了，区别就是读到key过期之后，loadFunc变成了storeFunc
2. 定时删除：定时扫描所有的key，这种场景检测的key很全，但是key多了之后性能吃不消
3. 定期删除：定期循环随机取一批key，检测他们的过期时间，直到这批key过期的比例小于某个值或者达到时间限制就结束。但是这种会有漏网之key

（无论上述哪种方式，close时候都要梭哈一把到db中）

综上，感觉第1种实现最容易也最能满足需求，因为不需要scan这种费时的操作，并且某些kv并不支持类似scan(limit)的操作。你觉得呢？

另外，对于这个模式还有一个疑问，过期执行持久化到db是否有必要，能否跳过过期检测，直接close

**Comment by flycash** [comment]
如果这样，你只需要在memory cache上设计一个 OnEvicted 的回调就可以了。因为write back，这个让用户自己去写。再定义个 LoadFunc 不是很有意义

**Comment by flycash** [comment]
我其实是希望所有的底层实现都加上类似这个 OnEvicted，或者说能加则加。目前来看，非内存的实现，很难支持这个特性

**Comment by flycash** [comment]
Please translate it to English.

And I think you can only support OnEvicted for memory cache. In my mind, all implementations should try to support this feature. But it may be very difficult.

In this MR, write back cache decorator looks not very necessary. As I mentioned above, after defining the Close function to memory cache, we can remove the write back decorator.

File: client/cache/memory.go [hunk 1]
```diff
@@ -186,6 +187,11 @@ func (bc *MemoryCache) StartAndGC(config string) error {
 	return nil
 }
 
+// RegisterEvictedFunc for write back cache to register onEvicted function
+func (bc *MemoryCache) RegisterEvictedFunc(fn func(key string, val any)) {
```
[the remark this answers is not in the document]
Comment (hookokoko) [reply client/cache/memory.go:191]: I have tried the Option pattern, but due to the limit of definition of Instance,
```go
type Instance func() Cache
```
it is hard to pass an onEvicted parameter.
If I define the instance like this, it looks inelegant...
 ```go
type Instance func(opt ...any) Cache
```
So, I don't want to support OnEvicted function. From our discussion, I think it is enough to define the Close function to memory cache

**Comment by hookokoko** [comment]
I have no good idea to solve the problem mention above, so closed temporarily...

**Pull request closed without merging**